### PR TITLE
DD-728: Retrieve actual files from Data Station

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes DD-728: Retrieve actual files from Data Station

# Description of changes

* the required default bundle name `hibernate` moved from easy_db to `verificationDatabase` as that one is used for all sub-commands.
* implemented the new sub-command.

# How to test

Though each sub-command uses either deasy or dataverse, both must be up and running due to the single configuration. 

* in the root of easy-dtap: `start-preprovisioned-box.py`
* in the root of dd-verify-file-migration: 
  * once: run `start-env.sh` and change `etc/config.yml`:
    * dataverse baseUrl: 'https://demo.archaeology.datastations.nl/'
    * the apiKey for dataverse and password for both databases
  * per test, with VPN to reach the demo server (each command line in another terminal, or the second command in Intellij by running DdVerifyFileMigrationApplication configured with the specified arguments):

        rm -rf data;start-hsqldb.sh
        start.sh load-from-dataverse [-d doi:10.17026/dans-zta-6jwm] etc/config.yml

* [x] Or build, deploy on deasy, adjust dataverse url/apiKey configuration as above and:

      dd-verify-file-migration load-from-dataverse  [-d doi:10.17026/dans-zta-6jwm] /etc/opt/dans.knaw.nl/dd-verify-file-migration/config.yml > verify.log

  You may want to switch off info logging for `org.hibernate.engine.internal.StatisticalLoggingSessionEventListener`

# Related PRs

* [x] [dd-dans-dataverse-client-lib](https://github.com/DANS-KNAW/dans-dataverse-client-lib/pull/14) needs to support  `DataFile.original*` for [10.17026/dans-xtz-qa6j](https://demo.archaeology.datastations.nl/api/datasets/export?exporter=dataverse_json&persistentId=doi%3A10.17026/dans-xtz-qa6j) merged, but no new version yet.

# Notify
@DANS-KNAW/dataversedans

### CODE REVIEW

File: src/main/java/nl/knaw/dans/filemigration/api/ActualFile.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  public int getVersion_nr() {
+    return version_nr;
+  }
+
+  public void setVersion_nr(int version_nr) {
+    this.version_nr = version_nr;
+  }
+
+  public void incRemoved_duplicate_file_count() {
+    this.version_nr += 1;
```
Comment (lindareijnhoudt) [inline src/main/java/nl/knaw/dans/filemigration/api/ActualFile.java]: what's this?

Comment (jo-pol) [reply src/main/java/nl/knaw/dans/filemigration/api/ActualFile.java]: used on retries for duplicates https://github.com/DANS-KNAW/dd-verify-file-migration/blob/e3e5b9d23a2d51a54ef724eae661000dae966fbd/src/main/java/nl/knaw/dans/filemigration/core/EasyFileLoader.java#L76

File: src/main/java/nl/knaw/dans/filemigration/core/DataverseLoader.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  public void saveActual(ActualFile actual) {
+    log.debug(actual.toString());
+    actualFileDAO.create(actual);
+  }
+
+  public void loadFromDataset(String doi) {
+    log.info("Loading {}", doi);
+    List<DatasetVersion> versions;
+    try {
+      versions = client.dataset(doi).getAllVersions().getData();
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/filemigration/core/DataverseLoader.java:55]: Stumbles on https://demo.archaeology.datastations.nl/api/datasets/export?exporter=dataverse_json&persistentId=doi%3A10.17026/dans-xtz-qa6j

with

    java.lang.RuntimeException: com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "originalFileFormat" (class nl.knaw.dans.lib.dataverse.model.file.DataFile), not marked as ignorable (13 known properties: "rootDataFileId", "pidURL", "filename", "creationDate", "persistentId", "contentType", "previousDataFileId", "id", "description", "filesize", "checksum", "storageIdentifier", "embargo"])
     at [Source: (StringReader); line: 1, column: 29161] (through reference chain: nl.knaw.dans.lib.dataverse.model.DataverseEnvelope["data"]->java.util.ArrayList[0]->nl.knaw.dans.lib.dataverse.model.dataset.DatasetVersion["files"]->java.util.ArrayList[28]->nl.knaw.dans.lib.dataverse.model.file.FileMeta["dataFile"]->nl.knaw.dans.lib.dataverse.model.file.DataFile["originalFileFormat"])

No success in `DdVerifyFileMigrationApplication.initialize` with

    bootstrap.getObjectMapper().disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);

File: src/main/java/nl/knaw/dans/filemigration/cli/LoadFromDataverseCommand.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            .create(
+                DataverseLoaderImpl.class,
+                new Class[] { DataverseClient.class, ActualFileDAO.class },
+                new Object[] { client, new ActualFileDAO(verificationBundle.getSessionFactory()) }
+            );
+        String doi = namespace.getString("doi");
+        if (doi != null) proxy.loadFromDataset(doi);
+        else {
+            log.info("No DOI provided, loading all datasets");
+            Iterator<ResultItem> resultItems = client.search().iterator("datasetVersionId:1", datasetOption());
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/filemigration/cli/LoadFromDataverseCommand.java]: Dar delivers the (manualy created?) dataset of the base box, demo returns nothing:
https://dar.dans.knaw.nl/api/search?q=datasetVersionId:1&per_page=10&start=0&type=dataset
https://demo.archaeology.datastations.nl/api/search?q=datasetVersionId:1&per_page=10&start=0&type=dataset

Looks like I misinterpreted datasetVersionId, it is not only unique per dataset but apparently for the whole dataverse. So what is alsway there, unique per dataset and indexed by solr. Not sure [this](https://github.com/DANS-KNAW/dataverse/blob/58a9f330484e9d823cd3ffce88b4d266ceeb366b/conf/solr/4.6.0/schema.xml#L300) is the proper source.

File: src/main/java/nl/knaw/dans/filemigration/cli/LoadFromDataverseCommand.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        Spliterator<ResultItem> itemSpliterator = spliteratorUnknownSize(itemIterator, Spliterator.ORDERED);
+        Stream<ResultItem> itemStream = stream(itemSpliterator, false);
+        return itemStream.map(ri -> getGlobalId((DatasetResultItem) ri)).filter(Objects::nonNull);
+    }
+
+    private String getGlobalId(DatasetResultItem ri) {
+        log.debug("id={} versionId={} majorVersion={} minorVersion={} fileCount={}",ri.getGlobalId(), ri.getVersionId(), ri.getMajorVersion(), ri.getMinorVersion(),ri.getFileCount());
+        if (1 == ri.getMajorVersion() && 0 == ri.getMinorVersion())
+            return ri.getGlobalId(); // workaround for unique DOIs
+        else return null; // filtered by caller
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/filemigration/cli/LoadFromDataverseCommand.java]: The query only returns 2.1 of [10.17026/dans-xtz-qa6j](https://demo.archaeology.datastations.nl/api/datasets/export?exporter=dataverse_json&persistentId=doi%3A10.17026/dans-xtz-qa6j) one of the hand-picked fedora datasets.

After further tests it is not (yet?) which version(s) of a dataset are returned by the solr query.

File: src/main/java/nl/knaw/dans/filemigration/core/DataverseLoader.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+  private final ActualFileDAO actualFileDAO;
+  private final DataverseClient client;
+
+  public DataverseLoader(DataverseClient client, ActualFileDAO actualFileDAO) {
+    this.actualFileDAO = actualFileDAO;
+    this.client = client;
+  }
+
+  public void saveActual(ActualFile actual) {
```
Comment (jo-pol) [inline src/main/java/nl/knaw/dans/filemigration/core/DataverseLoader.java:45]: Transaction per file by Impl class. Maybe you want a transaction per dataset.

Comment (lindareijnhoudt) [reply src/main/java/nl/knaw/dans/filemigration/core/DataverseLoader.java:45]: per dataset makes sense. Though there does not seem to be a strong reason to use transactions at all. no inconsistency will occur by only partly filling this table

File: src/main/java/nl/knaw/dans/filemigration/api/ActualFile.java (hunk 1 above)
Comment (lindareijnhoudt) [reply src/main/java/nl/knaw/dans/filemigration/api/ActualFile.java]: i still don't understand. isn't this the version as retrieved from Dataverse? why incrementing it? I understand this for the Expected file (there you don't increment the version, right?). But for the actual, a duplicate should produce an error. how can there be duplicate files in Dataverse? let's not fake a version that is not actually in Dataverse

Comment (jo-pol) [reply src/main/java/nl/knaw/dans/filemigration/api/ActualFile.java]: copy-paste devil from expected

**Pull request merged**